## pcadapt for random1 10K SNPs
rm(list = ls())

#!/usr/bin/Rscript
setwd("~/Documents/malaria_vector_reserach/input_data/a.aegypti/pcadapt/test/random1_10K")

library("qvalue")
library("pcadapt")
library("ggplot2")
      
## PC adapt tutorials:
## https://bcm-uga.github.io/pcadapt/articles/pcadapt.html
## https://popgen.nescent.org/2016-01-26-SNP-selection.html#section-4-logistic-regression-linking-outliers-and-temperature
## Note: PCAdapt expects the incoming matrix to have samples in columns and loci in rows
      

geno<-read.table(file ="A.aegypti_Q30.DP5.caling_1_maf0.03_good_inds_random1_10K.012.table", header = FALSE)  ## 012 genotype file generated by 012 vcftools followed by Rscript
geno_good<-geno[,-1]
#geno_good_lfmm<-t(geno_good)

#write.table(geno_good_lfmm, file ="test.inputs.lfmm", col.names = FALSE, row.names= FALSE, quote = FALSE)
write.table(geno_good, file ="A.aegypti_Q30.DP5.caling_1_maf0.03_good_inds_random1_10K.pcadapt", col.names = FALSE, row.names= FALSE, quote = FALSE)

#path_to_file <- "/share/lanzarolab/users/shaghayegh/data/A.aegypti/good_ind_Q30.DP5.caling_1_maf0.03/downstream_analysis/pcAdapt/test.inputs.pcadapt"
#filename <- read.pcadapt(path_to_file, type = "lfmm")

path_to_file <- "/Users/shaghayeghsoudi/Documents/malaria_vector_reserach/input_data/a.aegypti/pcadapt/test/random1_10K/A.aegypti_Q30.DP5.caling_1_maf0.03_good_inds_random1_10K.pcadapt"
filename <- read.pcadapt(path_to_file, type = "pcadapt")

x <- pcadapt(input = filename, K = 10) 
##screeplot
pdf(file = "screeplot_random1.pdf")
plot(x, option = "screeplot")
dev.off()

## score plot
## load ind file
poplist<-read.table(file = "gea.sample.list.ordered.vcf.txt", header = FALSE)
poplist.int<-poplist[,2]
poplist.names<-poplist[,1]

### compute test statistics
x <- pcadapt(filename, K = 3)
pdf(file = "pca_1_2_K3_random1_10K.pdf")
plot(x, option = "scores",  pop=poplist[,2])
dev.off()

pdf(file = "qqplot_K3_random1_10K.pdf")
plot(x, option = "qqplot")
dev.off()

pdf(file = "hist_K3_random1_10K.pdf")
hist(x$pvalues, xlab = "p-values", main = NULL, breaks = 50, col = "orange")
dev.off()

pdf(file = "stats.distribution_K3_random1_10.pdf")
plot(x, option = "stat.distribution")
dev.off()


plot(x, option = "manhattan")

### accounting for LD

path_to_file <- "/Users/shaghayeghsoudi/Documents/malaria_vector_reserach/input_data/a.aegypti/pcadapt/test/random1_10K/A.aegypti_Q30.DP5.caling_1_maf0.03_good_inds_random1_10K.pcadapt"
matrix <- read.pcadapt(path_to_file, type = "pcadapt")
res <- pcadapt(matrix, K = 20)

pdf(file = "screeplot_K20_matrix_random1_10.pdf")
plot(res, option = "screeplot")
dev.off()
res <- pcadapt(matrix, K = 3)
plot(res)



##To evaluate if LD might be an issue for your dataset, we recommend to display the loadings (contributions of each SNP to the PC) and to evaluate if the loadings are clustered in a single or several genomic regions.
pdf(file = "loading_PCs_random1_10k.pdf", height = 20, width= 15)
par(mfrow = c(3, 1))
for (i in 1:3)
  plot(res$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))

dev.off()

### thin SNPS for LD
res <- pcadapt(matrix, K = 20, LD.clumping = list(size = 200, thr = 0.1))
plot(res, option = "screeplot")


#After SNP thinning, we choose k

res <- pcadapt(matrix, K = 2, LD.clumping = list(size = 200, thr = 0.1))
par(mfrow = c(1, 2))
for (i in 1:2)
  plot(res$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))


plot(res)

